core/src/agent/state.rs

```diff
@@ -98,62 +96,31 @@ impl AgentStateSnapshot {
     }
 }
 
-impl ToJson for AgentStateSnapshot {
-    fn to_json(&self) -> Result<String, HolochainError> {
-        Ok(serde_json::to_string(self)?)
-    }
-}
-
 impl AddressableContent for AgentStateSnapshot {
     fn content(&self) -> Content {
-        self.to_json()
-            .expect("could not Jsonify ChainHeader as Content")
+        self.to_owned().into()
     }
 
     fn from_content(content: &Content) -> Self {
-        AgentStateSnapshot::from_json_str(content)
-            .expect("could not read Json as valid ChainHeader Content")
+        Self::try_from(content.to_owned()).expect("could not deserialized content")
     }
 
     fn address(&self) -> Address {
-        Address::from("AgentState")
+        "AgentState".into()
     }
 }
 
-#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize, DefaultJson)]
 /// the agent's response to an action
 /// stored alongside the action in AgentState::actions to provide a state history that observers
 /// poll and retrieve
 // @TODO abstract this to a standard trait
 // @see https://github.com/holochain/holochain-rust/issues/196
 pub enum ActionResponse {
     Commit(Result<Address, HolochainError>),
-    GetEntry(Option<Entry>),
+    GetEntry(Option<SerializedEntry>),
     GetLinks(Result<Vec<Address>, HolochainError>),
-    LinkEntries(Result<Entry, HolochainError>),
-}
-
-impl ToJson for ActionResponse {
-    fn to_json(&self) -> Result<String, HolochainError> {
-        match self {
-            ActionResponse::Commit(result) => match result {
-                Ok(entry_address) => Ok(format!("{{\"address\":\"{}\"}}", entry_address)),
-                Err(err) => Ok((*err).to_json()?),
-            },
-            ActionResponse::GetEntry(result) => match result {
-                Some(entry) => Ok(entry.to_json()?),
-                None => Ok("".to_string()),
-            },
-            ActionResponse::GetLinks(result) => match result {
-                Ok(hash_list) => Ok(json!(hash_list).to_string()),
-                Err(err) => Ok((*err).to_json()?),
-            },
-            ActionResponse::LinkEntries(result) => match result {
-                Ok(entry) => Ok(format!("{{\"address\":\"{}\"}}", entry.address())),
-                Err(err) => Ok((*err).to_json()?),
-            },
-        }
-    }
+    LinkEntries(Result<SerializedEntry, HolochainError>),
```

Comment:
Why are you using SerializedEntry in those actions instead of just Entry, @thedavidmeister?

Reply:
I think this is explained in the PR documentation.  I think it's because of wanting to get at the entry type separate from the entry value and they are both inside the entry and thus get serialized twice.  Or something like that.  It's not totally clear to me.